APOC 3.5.0.3 changes import file URL paths too aggressively

## Expected Behavior (Mandatory)

```
call apoc.metrics.get("neo4j.bolt.sessions_started");
```

Should return metric data in Neo4j Desktop installations.

## Actual Behavior (Mandatory)

```
Neo.ClientError.Procedure.ProcedureCallFailed: Failed to invoke procedure `apoc.metrics.get`: 

Caused by: java.io.FileNotFoundException: 
/Users/davidallen/Library/Application Support/Neo4j Desktop/Application/neo4jDatabases/database-6df59434-8fe6-405c-ab17-dd0a913b1498/installation-3.5.4/import/Users/davidallen/Library/Application Support/Neo4j Desktop/Application/neo4jDatabases/database-6df59434-8fe6-405c-ab17-dd0a913b1498/installation-3.5.4/metrics/neo4j.bolt.sessions_started.csv 
(No such file or directory)
```

## How to Reproduce the Problem

(Run above query)

## Cause

In `apoc.metrics.Metrics.loadCsvForMetric`, great care is taken to build an absolute directory reference to the location of metrics, taking into account neo4j config.  This is done by `FileUtils.getMetricsDirectory`.    But when this absolute URL gets passed to the LoadCsv methods, it appears to use `FileUtils.changeFileUrlIfImportDirectoryConstrained` which seems to be appending the absolute directory as a relative to the import path.

The result is an ultra-long local URL which is wrong, causing the file not to exist, causing the metrics load to fail.

@conker84 this seems to have fixed the path bug but introduced a different one.   Now I get this:

```
Neo.ClientError.Procedure.ProcedureCallFailed: Failed to invoke procedure `apoc.metrics.get`: Caused by: java.io.IOException: Stream closed
```

You can reproduce by running:

```call apoc.metrics.get("neo4j.bolt.sessions_started");```

--- a/src/main/java/apoc/load/LoadCsv.java
+++ b/src/main/java/apoc/load/LoadCsv.java
@@ -22,12 +22,14 @@ import java.util.stream.Collectors;
 import java.util.stream.Stream;
 import java.util.stream.StreamSupport;
 
+import static apoc.load.util.LoadCsvConfig.DEFAULT_ARRAY_SEP;
+import static apoc.load.util.LoadCsvConfig.Results;
+import static apoc.util.FileUtils.closeReaderSafely;
 import static apoc.util.Util.cleanUrl;
 import static apoc.util.Util.parseCharFromConfig;
 import static java.util.Arrays.asList;
 import static java.util.Collections.emptyList;
 import static java.util.Collections.emptyMap;
-import static apoc.load.util.LoadCsvConfig.*;
 
 public class LoadCsv {
 
@@ -39,28 +41,33 @@ public class LoadCsv {
     @Description("apoc.load.csv('url',{config}) YIELD lineNo, list, map - load CSV fom URL as stream of values,\n config contains any of: {skip:1,limit:5,header:false,sep:'TAB',ignore:['tmp'],nullValues:['na'],arraySep:';',mapping:{years:{type:'int',arraySep:'-',array:false,name:'age',ignore:false}}")
     public Stream<CSVResult> csv(@Name("url") String url, @Name(value = "config",defaultValue = "{}") Map<String, Object> configMap) {
         LoadCsvConfig config = new LoadCsvConfig(configMap);
+        CountingReader reader = null;
         try {
-            CountingReader reader = FileUtils.readerFor(url);
-
-            // new CSVReader(...) is deprecated, moved to the new builder
-            CSVReader csv = new CSVReaderBuilder(reader)
-                    .withCSVParser(new CSVParserBuilder()
-                            .withQuoteChar(config.getQuoteChar())
-                            .withSeparator(config.getSeparator())
-                            .build())
-                    .build();
-
-            String[] header = getHeader(csv, config);
-            boolean checkIgnore = !config.getIgnore().isEmpty() || config.getMappings().values().stream().anyMatch( m -> m.ignore);
-            return StreamSupport.stream(new CSVSpliterator(csv, header, url, config.getSkip(), config.getLimit(),
-                    checkIgnore, config.getMappings(), config.getNullValues(), config.getResults(), config.getIgnoreErrors()), false);
+            reader = FileUtils.readerFor(url);
+            return streamCsv(url, config, reader);
         } catch (IOException e) {
-
+            closeReaderSafely(reader);
             if(!config.isFailOnError())
-                return Stream.of(new  CSVResult(new String[0], new String[0], 0, true, Collections.emptyMap(), emptyList(), EnumSet.noneOf(Results.class)));
+                return Stream.of(new CSVResult(new String[0], new String[0], 0, true, Collections.emptyMap(), emptyList(), EnumSet.noneOf(Results.class)));
             else
                 throw new RuntimeException("Can't read CSV from URL " + cleanUrl(url), e);
         }
+    }
+
+    public Stream<CSVResult> streamCsv(@Name("url") String url, LoadCsvConfig config, CountingReader reader) throws IOException {
+
+        CSVReader csv = new CSVReaderBuilder(reader)
+                .withCSVParser(new CSVParserBuilder()
+                        .withQuoteChar(config.getQuoteChar())
+                        .withSeparator(config.getSeparator())
+                        .build())
+                .build();
+
+        String[] header = getHeader(csv, config);
+        boolean checkIgnore = !config.getIgnore().isEmpty() || config.getMappings().values().stream().anyMatch(m -> m.ignore);
+        return StreamSupport.stream(new CSVSpliterator(csv, header, url, config.getSkip(), config.getLimit(),
+                checkIgnore, config.getMappings(), config.getNullValues(), config.getResults(), config.getIgnoreErrors()), false)
+                .onClose(() -> closeReaderSafely(reader));
     }
 
     public static class Mapping {

--- a/src/main/java/apoc/metrics/Metrics.java
+++ b/src/main/java/apoc/metrics/Metrics.java
@@ -1,7 +1,9 @@
 package apoc.metrics;
 
 import apoc.ApocConfiguration;
+import apoc.export.util.CountingReader;
 import apoc.load.LoadCsv;
+import apoc.load.util.LoadCsvConfig;
 import apoc.util.FileUtils;
 import apoc.util.Util;
 import org.neo4j.logging.Log;
@@ -14,6 +16,8 @@ import java.util.HashMap;
 import java.util.Map;
 import java.util.function.Predicate;
 import java.util.stream.Stream;
+
+import static apoc.util.FileUtils.closeReaderSafely;
 
 /**
  * @author moxious
@@ -162,10 +166,18 @@ public class Metrics {
                     "https://neo4j.com/docs/operations-manual/current/monitoring/metrics/expose/#metrics-csv");
         }
 
-        String url = new File(metricsDir, metricName + ".csv").toURI().toString();
-        return new LoadCsv().csv(url, config)
-                .filter(Metrics.duplicatedHeaderRows)
-                .map(csvResult -> new GenericMetric(metricName, Util.toLong(csvResult.map.get("t")), csvResult.map));
+        String url = new File(metricsDir, metricName + ".csv").getAbsolutePath();
+        CountingReader reader = null;
+        try {
+            reader = FileUtils.readFile(url);
+            return new LoadCsv()
+                    .streamCsv(url, new LoadCsvConfig(config), reader)
+                    .filter(Metrics.duplicatedHeaderRows)
+                    .map(csvResult -> new GenericMetric(metricName, Util.toLong(csvResult.map.get("t")), csvResult.map));
+        } catch (Exception e) {
+            closeReaderSafely(reader);
+            throw new RuntimeException(e);
+        }
     }
 
     @Procedure(mode=Mode.DBMS)

--- a/src/main/java/apoc/util/FileUtils.java
+++ b/src/main/java/apoc/util/FileUtils.java
@@ -73,7 +73,7 @@ public class FileUtils {
         }
     }
 
-    private static CountingReader readFile(String fileName) throws IOException, FileNotFoundException {
+    public static CountingReader readFile(String fileName) throws IOException, FileNotFoundException {
         File file = new File(fileName);
         if (!file.exists() || !file.isFile() || !file.canRead()) throw new IOException("Cannot open file "+fileName+" for reading.");
         return new CountingReader(file);
@@ -260,4 +260,10 @@ public class FileUtils {
             "dbms.directories.tx_log",
             "unsupported.dbms.directories.neo4j_home"
     );
+
+    public static void closeReaderSafely(CountingReader reader) {
+        if (reader != null) {
+            try { reader.close(); } catch (IOException ignored) { }
+        }
+    }
 }

--- a/src/test/java/apoc/metrics/MetricsTest.java
+++ b/src/test/java/apoc/metrics/MetricsTest.java
@@ -1,0 +1,69 @@
+package apoc.metrics;
+
+import apoc.util.Neo4jContainerExtension;
+import apoc.util.TestUtil;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.neo4j.driver.v1.Session;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static apoc.util.TestContainerUtil.*;
+import static apoc.util.Util.map;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assume.assumeNotNull;
+
+/**
+ * @author as
+ * @since 13.02.19
+ */
+public class MetricsTest {
+
+    private static Neo4jContainerExtension neo4jContainer;
+    private static Session session;
+
+    @BeforeClass
+    public static void beforeAll() {
+        TestUtil.ignoreException(() -> {
+            neo4jContainer = createEnterpriseDB(true)
+                .withNeo4jConfig("apoc.import.file.enabled", "true");
+            neo4jContainer.start();
+        }, Exception.class);
+        assumeNotNull(neo4jContainer);
+        session = neo4jContainer.getSession();
+    }
+
+    @AfterClass
+    public static void afterAll() {
+        if (neo4jContainer != null) {
+            neo4jContainer.close();
+        }
+        cleanBuild();
+    }
+
+    @Test
+    public void shouldGetMetrics() {
+        session.readTransaction(tx -> tx.run("RETURN 1 AS num;"));
+        String metricKey = "neo4j.bolt.sessions_started";
+        testResult(session, "CALL apoc.metrics.get({metricKey})",
+                map("metricKey", metricKey), (r) -> {
+                    assertTrue("should have at least one element", r.hasNext());
+                    Map<String, Object> map = r.next();
+                    assertEquals(Stream.of("timestamp", "metric", "map").collect(Collectors.toSet()), map.keySet());
+                    assertEquals(metricKey, map.get("metric"));
+                });
+    }
+
+    @Test
+    public void shouldListMetrics() {
+        testResult(session, "CALL apoc.metrics.list()",
+                Collections.emptyMap(),
+                (r) -> assertTrue("should have at least one element", r.hasNext()));
+    }
+
+}

